In [62]:
import json # Used to loan data set
import functools # Used to maintain introspection on decorators
import time # Used to time functions

from collections import OrderedDict

def functionTimer(func):
    """Prints the functions runtime"""
    @functools.wraps(func) # maintains introspection
    def wrapper(*args, **kwargs):
        start = time.perf_counter()
        func(*args, **kwargs)
        end = time.perf_counter()
        runtime = end - start
        print (f"Finished {func.__name__!r} in {runtime: .4f} secs")
        return func(*args, **kwargs)
    return wrapper

@functionTimer
def remove_duplicates():
    '''
    Does Somthing.
    
        Parameters:
        
        Returns:
    '''
    # Load data
    with open('org_names.json', mode='r') as f:
        data = json.load(f)

    original_size = len(data) # To use as a comparison
    print("Size of original data set:", original_size)

    # Convert list to a dictionary then back to a list to automatically remove identical items
    data = list(dict.fromkeys(data)) # Removes 5757
    
    final_size = len(data) # To use as a comparison
    print("Number of removed names:", original_size - final_size)
    #print(data)

remove_duplicates()

Size of original data set: 8282
Number of removed names: 5757
Finished 'remove_duplicates' in  0.0048 secs
Size of original data set: 8282
Number of removed names: 5757


In [11]:
import functools # used to maintain introspection on decorators
import time # used to time functions

def functionTimer(func):
    """Prints the functions runtime"""
    @functools.wraps(func) # maintains introspection
    def wrapper(*args, **kwargs):
        start = time.perf_counter()
        func(*args, **kwargs)
        end = time.perf_counter()
        runtime = end - start
        print (f"Finished {func.__name__!r} in {runtime: .4f} secs")
        return func(*args, **kwargs)
    return wrapper

# Convert list to a dictionary then back to a list to automatically remove identical items
data = list(dict.fromkeys(data)) 
# Removes 5757 
# In 0.0060
# In 0.0057
# In 0.0042
# Average = 0.0043

# Convert list to a set then back to a list to automatically remove identical items
data = list(set(data))
# Removes 5757 
# In 0.0072
# In 0.0079
# In 0.0048
# Average = 0.0054

# Convert list to a dictionary then back to a list using the collections moduel 
data = list(OrderedDict.fromkeys(data))
# Removes 5757 
# In 0.0057
# In 0.0078
# In 0.0088
# Average = 0.0059

TASK
Background: Every time one of our users runs a report our Search & NLP pipeline fetches loads of information about them from company records, news articles, and blog posts. After completing this it will be able to generate a list of all company names it’s seen, we want to deduplicate entries in this list, so that our Report Generation pipeline can group together information about the same company before displaying it to the user.
Task: Write a function which takes in a list of company/organisation names, and reduces the list to one entry per company. Lots of companies will be in their multiple times, possibly with different forms of their names. The file provided should be treated as a representative example of the list of names your function will receive. You should carefully consider the time complexity of your code.
Include unit tests to show that your function works on various edge cases. The problem is fairly ill-defined and open ended, so part of the challenge is to define some bounds for your work! We're not looking for a complete solution - just a decent stab, a few good ideas and some sensible code.
Feel free to import and use any libraries or external datasets you like - that's a key part of real-world software development.
You should take care to ensure your code is readable and well-commented where necessary. Share your code (the whole solution, including the tests) with us (zip in an email, or public GitHub project). Please also share with us any additional comments, thoughts, or ideas about any difficulties you encountered in this task, and please take care to explain any limitations of your solution which you're aware of.

STEPS
- Write a function
- Function takes in a list
- Reduce the list to 1 entry per company
- Consider time complexity
- Unit tests prooving it deals with edge cases
- Readable and well commented
- Share ideas and difficulties
- Explain limitations

IDEAS
Use a set or dictionary to remove duplicates automatically
Store the size of the list to compare after cleaning to see how many duplicates were removed
Use a decorator to time the function

Remove obvious words (without editing them) like 'LIMITED' and 'LTD'
Use wildcards to remove the rest
Do as few cycles of the list as possible to maintain speed
Refactor funtion into smaller functions?
Rewrite function timer to take an average of 3

QUESTIONS
Is transfering to a set less speedy than cycling through a list?
Is there a method for removing duplicates? (Pandas?)
Test a random section instead of the whole document to reduce time?
Would a generator be useful?
Is using a dictionary better than using a set?
Some of the items are in caps, does case matter?

ANSWERS
A set is slower than a dictionary

RESOURCES USED
https://www.w3schools.com/python/python_howto_remove_duplicates.asp